File: src/dere_plugins/dere_vault/scripts/config_reader.py
#!/usr/bin/env python3
"""Read dere configuration for personality and daemon settings."""

from __future__ import annotations

import sys
import tomllib
from pathlib import Path


def get_config_path() -> Path:
    """Get path to dere config file."""
    # Check for config in standard locations
    config_locations = [
        Path.home() / ".config" / "dere" / "config.toml",
        Path.home() / ".dere" / "config.toml",
    ]

    for loc in config_locations:
        if loc.is_file():
            return loc

    # Default to first location
    return config_locations[0]


def load_config() -> dict:
    """Load dere configuration.

    Returns empty dict if config doesn't exist.
    """
    config_path = get_config_path()

    if not config_path.is_file():
        return {}

    try:
        with open(config_path, "rb") as f:
            return tomllib.load(f)
    except (OSError, tomllib.TOMLDecodeError):
        return {}


def get_personality(vault_path: Path | None = None) -> str | None:
    """Get personality for current vault.

    Returns None if no personality configured.
    """
    config = load_config()

    if not config:
        return None

    # Check for vault-specific personality
    if vault_path:
        vault_configs = config.get("vaults", {})
        vault_config = vault_configs.get(str(vault_path), {})

        if "personality" in vault_config:
            return vault_config["personality"]

    # Check for global default personality
    return config.get("default_personality")


def get_daemon_url() -> str | None:
    """Get daemon URL from config.

    Returns None if not configured.
    """
    config = load_config()
    return config.get("daemon_url")


def is_daemon_enabled(vault_path: Path | None = None) -> bool:
    """Check if daemon integration is enabled for vault.

    Returns False if not configured.
    """
    config = load_config()

    if not config:
        return False

    # Check vault-specific setting
    if vault_path:
        vault_configs = config.get("vaults", {})
        vault_config = vault_configs.get(str(vault_path), {})

        if "enable_daemon" in vault_config:
            return vault_config["enable_daemon"]

    # Check global setting
    return config.get("enable_daemon", False)


if __name__ == "__main__":
    # CLI usage
    if len(sys.argv) > 1:
        command = sys.argv[1]

        if command == "personality":
            vault_path = Path(sys.argv[2]) if len(sys.argv) > 2 else None
            personality = get_personality(vault_path)
            if personality:
                print(personality)
                sys.exit(0)
            else:
                sys.exit(1)

        elif command == "daemon-url":
            url = get_daemon_url()
            if url:
                print(url)
                sys.exit(0)
            else:
                sys.exit(1)

        elif command == "daemon-enabled":
            vault_path = Path(sys.argv[2]) if len(sys.argv) > 2 else None
            enabled = is_daemon_enabled(vault_path)
            print("true" if enabled else "false")
            sys.exit(0 if enabled else 1)

    else:
        print(
            "Usage: config_reader.py [personality|daemon-url|daemon-enabled] [vault_path]",
            file=sys.stderr,
        )
        sys.exit(1)

File: plugins/dere_vault/scripts/config_reader.ts
import { parse } from "@iarna/toml";
import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import { homedir } from "node:os";
import { join } from "node:path";

export function getConfigPath(): string {
  const locations = [
    join(homedir(), ".config", "dere", "config.toml"),
    join(homedir(), ".dere", "config.toml"),
  ];
  for (const loc of locations) {
    if (existsSync(loc)) {
      return loc;
    }
  }
  return locations[0];
}

export async function loadConfig(): Promise<Record<string, unknown>> {
  const configPath = getConfigPath();
  if (!existsSync(configPath)) {
    return {};
  }
  try {
    const raw = await readFile(configPath, "utf-8");
    return (parse(raw) as Record<string, unknown>) ?? {};
  } catch {
    return {};
  }
}

export async function getPersonality(vaultPath?: string | null): Promise<string | null> {
  const config = await loadConfig();
  const vaults = (config.vaults ?? {}) as Record<string, any>;

  if (vaultPath && vaults[vaultPath]?.personality) {
    return String(vaults[vaultPath].personality);
  }

  if (config.default_personality) {
    return String(config.default_personality);
  }

  return null;
}

export async function getDaemonUrl(): Promise<string | null> {
  const config = await loadConfig();
  const url = config.daemon_url;
  return url ? String(url) : null;
}

export async function isDaemonEnabled(vaultPath?: string | null): Promise<boolean> {
  const config = await loadConfig();
  const vaults = (config.vaults ?? {}) as Record<string, any>;

  if (vaultPath && typeof vaults[vaultPath]?.enable_daemon === "boolean") {
    return vaults[vaultPath].enable_daemon;
  }

  return Boolean(config.enable_daemon);
}

async function main(): Promise<void> {
  const command = process.argv[2];
  if (!command) {
    console.error("Usage: config_reader.ts [personality|daemon-url|daemon-enabled] [vault_path]");
    process.exit(1);
  }

  if (command === "personality") {
    const vaultPath = process.argv[3] ?? null;
    const personality = await getPersonality(vaultPath);
    if (personality) {
      console.log(personality);
      process.exit(0);
    }
    process.exit(1);
  }

  if (command === "daemon-url") {
    const url = await getDaemonUrl();
    if (url) {
      console.log(url);
      process.exit(0);
    }
    process.exit(1);
  }

  if (command === "daemon-enabled") {
    const vaultPath = process.argv[3] ?? null;
    const enabled = await isDaemonEnabled(vaultPath);
    console.log(enabled ? "true" : "false");
    process.exit(enabled ? 0 : 1);
  }

  console.error(`Unknown config key: ${command}`);
  process.exit(1);
}

if (import.meta.main) {
  void main();
}
